rfcs/2021-08-29-8381-vrl-iteration-support.md

```diff
@@ -0,0 +1,756 @@
+call_.
+
+That is, we don't want to allow this:
+
+```coffee
+my_closure = |k, v| [k, v]
+map(., my_closure)
+```
+
+As that would require us to (1) allow for a closure to be represented as a `Value`,
+since everything assigned to a variable has to be representable as a value, and
+(2) it would require us to add compiler-support to reject assigning closures to
+an event field (e.g. `.foo = |k, v| [k, v]`), as that would lead to its own set
+of problems.
+
+While closures-as-values might seem useful at first, there hasn't been any
+use-case that has come up in the past that makes this a requirement.
+
+Instead, the closure-syntax is tied to a function call, and can only be added to
+functions that explicitly expose their ability to take a closure with `x`
+arguments that returns `y` value.
+
+#### Lexical Scoping
+
+Lexical scoping (variables being accessible within a given scope, instead of
+globally) is something we've discussed before.
+
+Before, we decided that the complexity of adding lexical scoping wasn't worth
+the investment before our first release, and we also hoped that lexical scoping
+wouldn't be something that was ever needed in VRL.
+
+With this feature, and particular the function-closure syntax, lexical scoping
+comes to top of mind again.
+
+The reason for that, is the following example:
+
+```coffee
+map(.) |key, value| {
+  key = upcase(key)
+
+  [key, value]
+}
+
+key
+```
+
+We reference `key` outside the closure, at the last line of the program. What
+should the value of `key` be in this case?
+
+Without lexical scoping, it would be set to the upper-case variant of the "last"
+key in the event.
+
+With lexical scoping, it would return an "undefined variable" error at
+compile-time, because the `key` variable _inside_ the closure is
+lexically-scoped to that block, and remains undefined outside of the block.
+
+However, while the above syntax would be _new_ and thus not a breaking change,
+for existing code, adding lexical scoping _would_ be a breaking change:
+
+```coffee
+foo = "bar"
+
+{
+  foo = "baz"
+}
+
+foo
+```
+
+Previously, `foo` would return `"baz"` when the program runs, but with lexical
+scoping, it would be set to `"bar"` instead (depending on if we also added
+let-bindings or not).
+
+Because of this breaking change, this RFC proposes we add lexical scoping
+selectively, function-closures _will_ have _partial_ lexical scoping, anything else
+_won't_.
+
+What this means for function-closures is best demonstrated as an example:
+
+```coffee
+value1 = 1
+
+map(.) |key, value2| {
+  value1 = 1.1
+  value3 = 3
+
+  [key, value2]
+}
+
+value1
+value2
+value3
+```
+
+In the above example, `value1` returns `1.1`, `value2` returns an "undefined
+variable" error at compile-time, and `value3` returns `3`. In other words, lexical-scoping
+_only_ applies to the provided closure variables, _not_ to any variables defined
+inside the closure itself.
```

Comment:
I would prefer, 

```
value1 = 1.1
value2 = undefined
value3 = undefined
```

As that is what I would expect from most programming languages. The var hoisting you get in javascript which is what you propose here with `value3` was pretty unpopular - for good reason.

I also think it would be more straighforward to implement as we just implement our variable list as a stack and consistent rules should appy when accessing the variables. Variables are accessed by checking for existence in the top level of the stack, if they don't exist there then check in the previous layer. If there is no previous layer, then error. When exiting the block we wouldn't need to concern ourselves with any copying values and types between layers of the stack.

The concern with the block parameter variables can be addressed by treating the parameters as new definitions, and thus new variables. Essentially, the block just pushes new definitions for these variable names to the scope stack, so accessing the variables types wouldn't fall through to the previous level.

Reply:
I agree with this.

Reply:
Same

Reply:
Agreed, I'll make the change.

Reply:
Adding to https://github.com/timberio/vector/pull/9001/files#r701814426, I'd also prefer `value1 = undefined` if we can make it.